### PR TITLE
[RFR] Integrate aor-graphql packages

### PR DESCRIPTION
**Pull request by djhi**

### CODE REVIEW

File: packages/ra-graphql-client-graphcool/README.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+import { Admin, Resource, Delete } from 'react-admin';
+
+import { PostCreate, PostEdit, PostList } from './posts';
+
+const client = new ApolloClient();
+
+class App extends Component {
+    constructor() {
+        super();
+        this.state = { restClient: null };
```
Comment (fzaninotto) [inline packages/ra-graphql-client-graphcool/README.md]: should be renamed to `dataProvider`

File: packages/ra-graphql-client-graphcool/README.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+```
+
+## Usage
+
+This example assumes a `Post` type is defined in the graphcool schema.
+
+```js
+// in App.js
+import React, { Component } from 'react';
+import buildApolloClient from 'ra-graphql-client-graphcool';
```
Comment (fzaninotto) [inline packages/ra-graphql-client-graphcool/README.md]: replace `buildApolloClient` by `buildApolloProvider`

File: packages/ra-graphql-client-graphcool/package.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "ra-graphql-client-graphcool",
+  "version": "2.0.0-alpha3",
+  "description": "A Graphcool GraphQL client for react-admin",
```
Comment (fzaninotto) [inline packages/ra-graphql-client-graphcool/package.json]: data provider

File: Makefile [hunk 4]
```diff
@@ -26,7 +26,15 @@ build-ra-input-rich-text:
 	@NODE_ENV=production ./node_modules/.bin/babel ./packages/ra-input-rich-text/src -d ./packages/ra-input-rich-text/lib --ignore '*.spec.js'
 	@cd packages/ra-input-rich-text/src && rsync -R `find . -name *.css` ../lib
 
-build: build-react-admin build-ra-data-json-server build-ra-data-simple-rest build-ra-input-rich-text ## compile ES6 files to JS
+build-ra-graphql-client:
```
Comment (fzaninotto) [inline Makefile]: data provider

File: packages/ra-graphql-client-graphcool/README.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
+# ra-graphql-client-graphcool
```
Comment (fzaninotto) [inline packages/ra-graphql-client-graphcool/README.md]: rename to `ra-data-graphcool`

File: packages/ra-graphql-client-graphcool/src/graphqlify/README.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# graphqlify
+
+graphqlify is a project from the now closed KadiraHQ company.
```
Comment (fzaninotto) [inline packages/ra-graphql-client-graphcool/src/graphqlify/README.md]: how about the license?

File: packages/ra-graphql-client/README.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
+# ra-graphql-client
```
Comment (fzaninotto) [inline packages/ra-graphql-client/README.md]: rename to `ra-data-graphql`

File: packages/ra-graphql-client/README.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+- [Installation](#installation)
+- [Usage](#installation)
+- [Options](#options)
+
+This is a very low level library which is not meant to be used directly unless you really want full control or are building a custom GraphQL client.
+
+It provides the foundations for other packages such as:
+
+- [ra-graphql-client-simple](https://github.com/marmelab/ra-graphql-client-simple)
```
Comment (fzaninotto) [inline packages/ra-graphql-client/README.md]: not part of this PR?

File: packages/ra-graphql-client/README.md [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+```sh
+yarn add ra-graphql-client
+```
+
+## Usage
+
+```jsx
+// in App.js
+import React, { Component } from 'react';
+import buildApolloClient from 'ra-graphql-client';
```
Comment (fzaninotto) [inline packages/ra-graphql-client/README.md]: data provider

File: packages/ra-graphql-client/README.md [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+You can pass any options supported by the [ApolloClient](http://dev.apollodata.com/core/apollo-client-api.html#apollo-client) contructor with the addition of `uri` which can be specified so that we create the network interface for you.
+
+You can also supply your own [ApolloClient](http://dev.apollodata.com/core/apollo-client-api.html#apollo-client) instance directly with:
+
+```js
+buildApolloClient({ client: myClient });
+```
+
+### IntrospectionOptions
```
Comment (fzaninotto) [inline packages/ra-graphql-client/README.md]: s/IntrospectionOptions/Introspection Options/

File: packages/ra-graphql-client/package.json [hunk 11]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "ra-graphql-client",
+  "version": "2.0.0-alpha3",
+  "description": "A GraphQL client for react-admin",
```
Comment (fzaninotto) [inline packages/ra-graphql-client/package.json]: data provider

File: packages/ra-graphql-client/README.md (hunk 8 above)
Comment (djhi) [reply packages/ra-graphql-client/README.md]: Never implemented. I'll remove it

File: packages/ra-data-graphcool/src/index.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import merge from 'lodash.merge';
+import buildClient from 'ra-data-graphql';
```
Comment (fzaninotto) [inline packages/ra-data-graphcool/src/index.js]: buildProvider?

File: packages/ra-data-graphql/README.md [hunk 13]
```diff
@@ -0,0 +1,230 @@
+# ra-data-graphql
+
+[![Build Status](https://travis-ci.org/marmelab/ra-data-graphql.svg?branch=master)](https://travis-ci.org/marmelab/ra-data-graphql)
+
+A GraphQL client for [react-admin](https://github.com/marmelab/react-admin/)
```
Comment (fzaninotto) [inline packages/ra-data-graphql/README.md]: data provider

File: packages/ra-data-graphql/README.md [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+built with [Apollo](http://www.apollodata.com/)
+
+A version of the `react-admin` demo using this client is available at https://marmelab.com/react-admin-graphql-demo.<br>
```
Comment (fzaninotto) [inline packages/ra-data-graphql/README.md]: data provider

File: packages/ra-data-graphql/README.md [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+
+- [Installation](#installation)
+- [Usage](#installation)
+- [Options](#options)
+
+This is a very low level library which is not meant to be used directly unless you really want full control or are building a custom GraphQL client.
```
Comment (fzaninotto) [inline packages/ra-data-graphql/README.md]: s/client/data provider/

File: packages/ra-data-graphql/README.md [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+});
+```
+
+The `./schema` file is a `schema.json` in `./scr` retrieved with [`get-graphql-schema --json <graphql_endpoint>`](https://github.com/graphcool/get-graphql-schema).
+
+> Note: Importing the `schema.json` file will significantly increase the bundle size.
+
+## Specify your queries and mutations
+
+For the client to know how to map react-admin request to apollo queries and mutations, you must provide a `queryBuilder` option. The `queryBuilder` is a factory function which will be called with the introspection query result.
```
Comment (fzaninotto) [inline packages/ra-data-graphql/README.md]: s/client/provider/

File: packages/ra-data-graphql/README.md [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+                ],
+                type : { kind: 'OBJECT', name: 'Category' }
+            },
+            ...
+        }
+    ]
+}
+```
+
+The `queryBuilder` function must return a function which will be called with the same parameters as the react-admin client but must return an object matching the `options` of the ApolloClient [query](http://dev.apollodata.com/core/apollo-client-api.html#ApolloClient.query) method with an additional `parseResponse` function.
```
Comment (fzaninotto) [inline packages/ra-data-graphql/README.md]: data provider

File: packages/ra-data-graphql/README.md [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+
+```js
+buildGraphQLProvider({ queryBuilder });
+```
+
+## Troubleshooting
+
+*When I create or edit a resource, the list or edit page does not refresh its data*
+
+`react-admin` maintain its own cache of resources data but, by default, so does the Apollo client. For every queries, we inject a default [`fetchPolicy`](http://dev.apollodata.com/react/api-queries.html#graphql-config-options-fetchPolicy) set to `network-only` so that the Apollo client always refetch the data when requested.
```
Comment (fzaninotto) [inline packages/ra-data-graphql/README.md]: data provider

File: packages/ra-data-graphql/src/buildApolloClient.js [hunk 19]
```diff
@@ -0,0 +1,18 @@
+import { ApolloClient, createNetworkInterface } from 'apollo-client';
```
Comment (fzaninotto) [inline packages/ra-data-graphql/src/buildApolloClient.js]: rename to `buildApolloProvider`

File: packages/ra-data-graphql/src/index.js [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+    GET_LIST,
+    GET_ONE,
+    GET_MANY,
+    GET_MANY_REFERENCE,
+    CREATE,
+    UPDATE,
+    DELETE,
+} from 'react-admin';
+
+import buildApolloClient from './buildApolloClient';
```
Comment (fzaninotto) [inline packages/ra-data-graphql/src/index.js]: `buildApolloProvider`

File: packages/ra-data-graphql/src/index.js [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+    if (introspection) {
+        introspectionResults = await resolveIntrospection(
+            client,
+            introspection
+        );
+    }
+
+    const buildQuery = buildQueryFactory(introspectionResults, otherOptions);
+
+    const aorClient = (aorFetchType, resource, params) => {
```
Comment (fzaninotto) [inline packages/ra-data-graphql/src/index.js]: aorClient => raProvider

**Comment by mrcasablr** [comment]
I hope this works with non-graphcool endpoints?

File: packages/ra-data-graphql/README.md (hunk 18 above)
Comment (djhi) [reply packages/ra-data-graphql/README.md]: Not this time: we are talking about the Apollo client, not the GraphQL dataProvider which uses this client

File: packages/ra-data-graphql/src/buildApolloClient.js (hunk 19 above)
Comment (djhi) [reply packages/ra-data-graphql/src/buildApolloClient.js]: Nope, it builds an Apollo client which will be used by the data provider

File: packages/ra-data-graphql/src/index.js (hunk 20 above)
Comment (djhi) [reply packages/ra-data-graphql/src/index.js]: Nope, it builds an Apollo client which will be used by the data provider

**Comment by djhi** [comment]
@mrcasablr yes, the `ra-data-graphql` provides the basis for any graphql flavor. It requires more work on your side though as we can't predict how you implemented your queries/mutations. That's why we only provide a complete client for graphcool as their API is known and documented.
